src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -97,29 +95,27 @@ public TopicPartitionChannel(
 
     this.streamingBuffer = new StreamingBuffer();
     this.processedOffset = new AtomicLong(-1);
-    this.committedOffset = new AtomicLong(0);
+    this.offsetSafeToCommitToKafka = new AtomicLong(0);
   }
 
   /* Update the channel */
   protected void updateStreamingIngestChannel(SnowflakeStreamingIngestChannel updatedChannel) {
     this.channel = updatedChannel;
   }
 
-  // inserts the record into buffer
+  /**
+   * Inserts the record into buffer
+   *
+   * <p>Step 1: Initializes this channel by fetching the offsetToken from Snowflake for the first
+   * time this channel/partition has received offset after start/restart.
+   *
+   * <p>Step 2: Decides whether given offset from Kafka needs to be processed and whether it
+   * qualifies for being added into buffer.
+   *
+   * @param record input record from Kafka
+   */
   public void insertRecordToBuffer(SinkRecord record) {
-    if (!hasChannelReceivedAnyRecordsBefore) {
-      LOGGER.info(
-          "Received offset:{} for topic:{} as the first offset for this partition:{} after"
-              + " start/restart",
-          record.kafkaOffset(),
-          record.topic(),
-          record.kafkaPartition());
-      // This will only be called once at the beginning when an offset arrives for first time
-      // after connector starts
-      fetchLastCommittedOffsetToken();
-
-      this.hasChannelReceivedAnyRecordsBefore = true;
-    }
+    initChannel(record);
 
     // discard the record if the record offset is smaller or equal to server side offset, or if
```

Comment:
> discard the record if the record offset is smaller or equal to server side offset

Is this possible to happen? If no, could you add a logwarn?

Reply:
Yess. if we had not committed that offset, upon restart it might send same offset again.

Reply:
Yes, good idea to add logwarn. might be helpful while debugging. I will also print processoffset and committedoffset from snowflaake.

Reply:
if this is a valid case, then no need to add logwarn